### PR TITLE
fixed bug in dask.array.roll() for roll-shifts the size of the input array

### PR DESCRIPTION
**Pull request by ParticularMiner**
- [x] Closes (not applicable)
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

Discovered a bug while working with `dask.array.roll()`:

```python
x = da.arange(2, 3)
y = da.roll(x, 1)
y[0] = 0
x.compute(), y.compute()
```

Output:
```python
(array([0]), array([0]))
```

Should have been:
```python
(array([2]), array([0]))
```

### CODE REVIEW

File: dask/array/tests/test_routines.py [hunk 1]
```diff
@@ -1329,6 +1329,14 @@ def test_roll(chunks, shift, axis):
         assert_eq(np.roll(x, shift, axis), da.roll(a, shift, axis))
 
 
+def test_roll_bug():
```
Comment (jcrist) [inline dask/array/tests/test_routines.py:1332]: Can you change this to a more specific test name? `test_roll_bug` doesn't mean much to a reader, but `test_roll_always_results_in_a_new_array` does mean something.

```suggestion
def test_roll_always_results_in_a_new_array():
```

File: dask/array/routines.py [hunk 2]
```diff
@@ -1865,6 +1865,7 @@ def roll(array, shift, axis=None):
         result = concatenate([result[sl1], result[sl2]], axis=i)
 
     result = result.reshape(array.shape)
+    result = result.copy() if result is array else result
```
Comment (jcrist) [inline dask/array/routines.py:1868]:
```suggestion
    # Ensure that the output is always a new array object
    result = result.copy() if result is array else result
```

File: dask/array/tests/test_routines.py [hunk 3]
```diff
@@ -1329,6 +1329,14 @@ def test_roll(chunks, shift, axis):
...
+    # This bug was exposed in GitHub Issue #8723
```
Comment (jcrist) [inline dask/array/tests/test_routines.py:1333]: No need for this comment, git blame already provides this information.

```suggestion
```

File: dask/array/tests/test_routines.py [hunk 4]
```diff
@@ -1329,6 +1329,14 @@ def test_roll(chunks, shift, axis):
...
+    x = da.arange(2, 3)
+    y = da.roll(x, 1)
+    y[0] = 0
+    assert x[0].compute() == 2
```
Comment (jcrist) [inline dask/array/tests/test_routines.py:1337]: It'd be simpler to assert what we're actually testing.

```suggestion
    assert y is not x
```

**Review by jcrist — COMMENTED** [review]
Thanks for finding this bug @ParticularMiner.

File: dask/array/tests/test_routines.py (hunk 1 above)
Comment (ParticularMiner) [reply dask/array/tests/test_routines.py:1332]: Sure.

**Comment by jcrist** [comment]
Also - sidenote - it looks like you've locked your repo for maintainers to make changes to your PRs (e.g. I can't apply my suggestions above myself). If this is intentional that's fine, but if not, in the future disabling this would let us make small fixes like this without requiring you to make those changes yourself.

File: dask/array/routines.py (hunk 2 above)
Comment (ParticularMiner) [reply dask/array/routines.py:1868]: Done.

**Review by jcrist — APPROVED** [review]

**Pull request merged**